Introducing virtual FE

ranj063 opened the pull request:
This set of patches introduces the concept of a virtual FE dai link.
Previously, when trying to enable the hostless pipeline, it was
discovered that the codec does not get enabled and the tone remained inaudible.
Therefore, in such cases, a virtual FE dai link will be used to establish
a connection to the BE dai and enable the codec when the pipeline is triggered.

The virtual FE will be created when a siggen widget is loaded and freed when
the siggen widget is unloaded. It will be used for enabling the BE dai in the
kcontrol IO handler for the kcontrol attached to the siggen by calling
the soc_spcm_runtime_update() method.

lgirdwood commented on include/sound/soc.h:
```diff
@@ -1134,6 +1134,17 @@ struct snd_soc_dai_link {
+	 * This connection will be established at runtime by triggering the
+	 * hostless pipeline with a kcontrol attached to the component.
+	 */
+	unsigned int virtual:1;
```
Just wondering if we can set dynamic = 1 and no_pcm = 1 to achieve the same effect and hence no need for this ?

plbossart replied:
i had the same question, it feels somewhat redundant with existing mechanisms

ranj063 replied:
Fixed in the revised version

lgirdwood commented on sound/soc/soc-pcm.c:
```diff
@@ -2789,6 +2788,114 @@ int soc_dpcm_be_digital_mute(struct snd_soc_pcm_runtime *fe, int mute)
+	/* define virtual FE DAI link */
+	link->virtual = 1;
+	link->name = link_name;
+	link->id = 1;
```
Pass ID in ?

ranj063 replied:
I removed the ID as it isnt really needed

lgirdwood commented on sound/soc/soc-pcm.c:
```diff
@@ -2789,6 +2788,114 @@ int soc_dpcm_be_digital_mute(struct snd_soc_pcm_runtime *fe, int mute)
+	link->platform_name = platform_name;
+	link->codec_name = "snd-soc-dummy";
+	link->codec_dai_name = "snd-soc-dummy-dai";
+	link->num_codecs = 1;
```
Do we need codecs ? should it b 0 ?

ranj063 replied:
Fixed

lgirdwood commented on sound/soc/soc-pcm.c:
```diff
@@ -2789,6 +2788,114 @@ int soc_dpcm_be_digital_mute(struct snd_soc_pcm_runtime *fe, int mute)
+	link->dynamic = 1;
+
+	/*TODO: check if we need to handle capture for virtual FE */
+	link->dpcm_playback = 1;
```
capture ?

plbossart replied:
yes, this looks like a generic capability on capture where a component can capture audio and throw some sort of non-audio event DAPM knows nothing about.

ranj063 replied:
Should the virtual FE have both dpcm_playback and dpcm_capture enabled by default ?
Or does it depend on the widget that calls vfe_new(). So in the case of tone, I can only have dpcm_playback enabled?

ranj063 replied:
I've added capture support in virtual FE too. The widget creating the vfe must specify the stream direction.

lgirdwood commented on sound/soc/soc-pcm.c:
```diff
@@ -2789,6 +2788,114 @@ int soc_dpcm_be_digital_mute(struct snd_soc_pcm_runtime *fe, int mute)
+EXPORT_SYMBOL_GPL(soc_dpcm_vfe_new);
+
+/* free virtual FE DAI link */
+int soc_dpcm_vfe_free(struct snd_soc_card *card)
```
we need to individually delete these link by link.

ranj063 replied:
Is the idea then that the widget that created the virtual fe, will delete it?
In that case, can I associate the link with the comp by adding it to the snd_sof_control associated with the comp?

ranj063 replied:
FIxed to delete vfe by name

lgirdwood commented on sound/soc/soc-pcm.c:
```diff
@@ -2789,6 +2788,114 @@ int soc_dpcm_be_digital_mute(struct snd_soc_pcm_runtime *fe, int mute)
+			for_each_rtdcom_safe(rtd, rtdcom1, rtdcom2)
+				kfree(rtdcom1);
+
+			INIT_LIST_HEAD(&rtd->component_list);
```
Why is the list head init here in free() ?

ranj063 replied:
fixed

lgirdwood commented on sound/soc/soc-dapm.c:
```diff
@@ -1246,7 +1246,8 @@ int snd_soc_dapm_dai_get_connected_widgets(struct snd_soc_dai *dai, int stream,
 
 	/* Drop starting point */
-	list_del(widgets.next);
+	if (!list_is_singular(&widgets))
```
What happens if you have more than one widget between widget and BE, it's valid and could be constructed.

ranj063 replied:
This part is a bit unclear to me. The problem I was facing was that the snd_soc_dapm_dai_get_connected_widgets() returned a list with only the BE DAI in the case of the tone pipeline and dropping the starting point resulted in an empty list of paths to process. 

the dai argument passed to this method was the cpu_dai->playback_widget() which is the BE dai in my case. And this was the only way I could get it to be enabled. I could use some help here to figure out the right way to do it.

plbossart replied:
Yes, what happens if you have tone followed by a volume control?

ranj063 replied:
We do have a tone followed by volume in my topology. But the snd_soc_dapm_dai_get_connected_widgets() is called starting with the dai->playback_widget as the starting point. And this is the BE DAI.

lgirdwood replied:
I think we should also state in a comment what we are doing here and why.

lgirdwood commented on sound/soc/soc-pcm.c:
```diff
@@ -3150,6 +3152,44 @@ int soc_new_pcm(struct snd_soc_pcm_runtime *rtd, int num)
+		 * pcm device and substream for the requested direction
+		 */
+		if (rtd->dai_link->virtual) {
+			struct snd_pcm_str *pstr;
```
I would offload this logic as a static function.

lgirdwood replied:
and can any part of this be done by call core ALSA functions directly ?

ranj063 replied:
The only function that does this is snd_new_pcm() but that also registers the pcm device. So I created this to avoid that part.

ranj063 commented:
Thanks, Liam and Pierre. Let me work on revising the patchset today.

ranj063 commented:
@plbossart , @lgirdwood . I've pushed some changes. Could you please review the new version?

plbossart commented on sound/soc/soc-pcm.c:
```diff
@@ -2789,6 +2788,109 @@ int soc_dpcm_be_digital_mute(struct snd_soc_pcm_runtime *fe, int mute)
+			dpcm_be_disconnect(rtd, stream_dir);
+
+			/* free pcm runtime */
+			kfree(rtd->dpcm[stream_dir].runtime);
```
Question: all these frees don't reflect what is done for the _new method, could we rely on a helper to avoid tracking all these actions? It looks like either copy-paste or something extracted from somewhere else.

ranj063 replied:
This is actually not a copy-paste at all. I traced all the steps from the point of adding the link to the time when the sound card is registered and freed all the data that is allocated. Here're the steps I traced:
1. Bind DAI link: The PCM runtime and the dpcm runtime is created during this step. 
There are some other runtime child structures allocated during this step too ie codec dais and component list. 
2. Probe DAI link :The pcm device and the substream is created here.
3. dpcm runtime update: the FE link is connected to the BE

So, in the vfe_free() method, I undid all of this. 

But I wonder if I can avoid all of this manually in the free method. Let me dig into these steps a bit more.

plbossart commented on sound/soc/soc-core.c:
```diff
@@ -1134,6 +1135,31 @@ static int soc_bind_dai_link(struct snd_soc_card *card,
 
 	soc_add_pcm_runtime(card, rtd);
+
+	/* if this is a virtual FE link, create runtime and set it as active */
```
I don't know what 'active' runtime means. I would think it becomes 'active' when the switch is On, but maybe it's a different concept you are referring to.

ranj063 replied:
Oh, what I mean by active here is when the kcontrol is switched on and the dpcm_runtime_update() is called, it checks to see if playback or capture is active on the FE link. So I set the FE playback/capture active status by default.

plbossart replied:
I see, but I wonder if this has negative implications related to suspend/resume? If there is nothing going on it's a bit odd to force a runtime to be active.

ranj063 replied:
Yes, thats a good point.

The problem is the runtime_update() method is oblivious to where it is being called from. It just looks for an active link to process the update. 

Maybe I can avoid this by activating the FE in the kcontrol IO handler just before calling runtime_update.

lgirdwood commented on sound/soc/soc-pcm.c:
```diff
@@ -1582,7 +1582,6 @@ static int dpcm_add_paths(struct snd_soc_pcm_runtime *fe, int stream,
 
 	/* Create any new FE <--> BE connections */
 	for (i = 0; i < list->num_widgets; i++) {
-
```
best to keep this line  as is.

lgirdwood commented on sound/soc/soc-pcm.c:
```diff
@@ -2996,6 +3100,55 @@ static int soc_rtdcom_mmap(struct snd_pcm_substream *substream,
+	substream->pstr = pstr;
+	substream->number = 0;
+	substream->stream = stream_dir;
+	sprintf(substream->name, "subdevice #%i", 0);
```
Is the name exposed by /sysff or /proc ? if so, do we have any naming collisions when we register 2 or more VFE ?

ranj063 replied:
hmm I did not check if it is exposed. Let me try to make it unique

plbossart commented on sound/soc/soc-dapm.c:
```diff
@@ -855,6 +855,7 @@ static int dapm_create_or_share_kcontrol(struct snd_soc_dapm_widget *w,
 			kcname_in_long_name = true;
 		} else {
 			switch (w->id) {
+			case snd_soc_dapm_siggen:
```
as discussed last week, I wonder if we should also add a snd_soc_dapm_sig_sink (the dual of a generator). This would be *very* useful for testing e.g. by letting us set-up all sort of DMA channels with a self-test checking.

ranj063 replied:
Sure, I will add it to my next update.

ranj063 commented:
@plbossart , @lgirdwood . I'm closing this pull request and replacing it with a new pull request that has all the commits to enable the tone generator. This will be helpful for you review the entire flow for enabling the VFE link and the pipeline. I have addressed your comments in the new pull request.
